tools/flaky_test_monitor/level2/process_summary2_results.go

```diff
@@ -0,0 +1,140 @@
+
+				// increment # of passes, fails, skips or no results for this test
+				testResultSummary.Runs++
+				switch testResult.Result {
+				case "pass":
+					testResultSummary.Passed++
+				case "fail":
+					testResultSummary.Failed++
+					saveFailureMessage(testResult)
+				case "skip":
+					testResultSummary.Skipped++
+
+					// don't count skip as a run
+					testResultSummary.Runs--
+
+					// truncate last duration - don't count durations of skips
+					testResultSummary.Durations = testResultSummary.Durations[:len(testResultSummary.Durations)-1]
+				case "":
+					testResultSummary.NoResult++
+				default:
+					panic(fmt.Sprintf("unexpected test result: %s", testResult.Result))
+				}
+
+				testSummary2.TestResults[mapKey] = testResultSummary
+			}
+		}
+	}
+
+	// calculate averages and other calculations that can only be completed after all test runs have been read
+	postProcessTestSummary2(testSummary2)
+	return testSummary2
+}
+
+func saveFailureMessage(testResult helpers.TestResult) {
+	if testResult.Result != "fail" {
+		panic(fmt.Sprintf("unexpected test result: " + testResult.Result))
+	}
+
+	// sub-directory names should be the same - each sub directory corresponds to a failed test name
+	// there could already be previous failures for this test, so it's important to check if failed
+	// test folder exists
+	if !helpers.FolderExists(failuresDir + testResult.Test) {
+		err := os.Mkdir(failuresDir+testResult.Test, 0755)
+		helpers.AssertNoError(err, "error creating sub-dir under failuresDir")
+	}
+
+	// under each sub-directory, there should be 1 or more text files (failure1.txt, failure2.txt, etc)
+	// that holds the raw failure message for that test
+
+	dirEntries, err := os.ReadDir(failuresDir + testResult.Test)
+	helpers.AssertNoError(err, "error reading sub-dir entries under failuresDir")
+
+	// failure text files will be named "failure1.txt", "failure2.txt", etc so we want to know how
+	// many failure files already exist in the sub-directory before creating the next one
+	failureFile, err := os.Create(failuresDir + testResult.Test + "/" + fmt.Sprintf("failure%d.txt", len(dirEntries)+1))
+	helpers.AssertNoError(err, "error creating failure file")
+
+	for _, output := range testResult.Output {
+		_, err = failureFile.WriteString(output)
+		helpers.AssertNoError(err, "error writing to failure file")
+	}
+}
+
+func postProcessTestSummary2(testSummary2 helpers.TestSummary2) {
+	for _, testResultSummary := range testSummary2.TestResults {
+		// calculate average duration for each test summary
+		var durationSum float32 = 0
+		for _, duration := range testResultSummary.Durations {
+			durationSum += duration
+		}
+		testResultSummary.AverageDuration = helpers.ConvertTo2DecimalPlaces2(durationSum, testResultSummary.Runs)
+
+		// calculate failure rate for each test summary
+		testResultSummary.FailureRate = helpers.ConvertTo2DecimalPlaces(testResultSummary.Failed, testResultSummary.Runs)
+	}
+}
+
+func main() {
```

Comment:
This is just a test run? I couldn't really tell from the file name, could we make it more clear?

Reply:
It's not a test, it's the `main()` function for level 2 - I added documentation for this `main()` and for the level 1 `main()` so their purpose is more clear.

https://github.com/onflow/flow-go/pull/1707/commits/ebd163c35b9499bb10d45698de196e18f21a6e40